Testscripts/Linux/XDP-Action.sh

```diff
@@ -0,0 +1,291 @@
++       void *data = (void *)(long)ctx->data;
++       struct ethhdr *eth = data;
++       struct iphdr *iph = (struct iphdr *)(eth + 1);
++       struct icmphdr *icmph = (struct icmphdr *)(iph + 1);
++
++       if (icmph + 1 > data_end || icmph->type == 0)
++               return XDP_PASS;
++
++       swap_mem(eth->h_dest, eth->h_source, ETH_ALEN);
++
++       iph->id = 0;
++       iph->frag_off = 0;
++       swap_mem(&iph->saddr, &iph->daddr, 4);
++       ip_csum(iph);
++
++       icmph->type = 0;
++       icmph->checksum += 8;
++
++       return XDP_TX;
++}
++
+ SEC("xdp")
+ int process_packet(struct xdp_md *ctx)
+ {
+@@ -136,6 +203,10 @@ int process_packet(struct xdp_md *ctx)
+	bpf_perf_event_output(ctx, &perf_map,
+			      (__u64)pkt.pkt_len << 32 | BPF_F_CURRENT_CPU,
+			      &pkt, sizeof(pkt));
++
++	if (pkt.l3_proto == ETH_P_IP && pkt.l4_proto == IPPROTO_ICMP)
++		return ping_reply(ctx);
++
+ 	return XDP_PASS;
+ }
+
+--
+2.17.1
+
+
+EOF
+	fi
+	if [[ ! -f /root/xdpdump_kern_"${ACTION}".patch ]]; then
+		LogErr "patch file '/root/xdpdump_kern_${ACTION}.patch' could not be created"
+		SetTestStateAborted
+		exit 1
+	fi
+	LogMsg "patch file for ${ACTION} is created successfully"
+}
+
+# Helper function
+# This function will start xdpdump on client 
+# and ping from server on eth1 network interface
+function ping_test () {
+	local logSuffix=$1
+
+	xdpdumpCommand="cd bpf-samples/xdpdump && timeout 20 ./xdpdump -i ${nicName} > ~/xdpdumpout_${logSuffix}.txt"
+	LogMsg "Starting xdpdump on client with command: ${xdpdumpCommand}"
+	ssh ${client} "sh -c '${xdpdumpCommand} &'"
+
+	LogMsg "Starting ping on server with command: ping -I ${nicName} -c 10 ${clientSecondIP} > ~/pingOut_${logSuffix}.txt"
+	ssh ${server} "ping -I ${nicName} -c 10 ${clientSecondIP} > ~/pingOut_${logSuffix}.txt"
+}
+
+# Helper function
+# This function will start tcpdump on client and start ping_test
+function ping_with_tcpdump () {
+	local logSuffix=$1
+	LogMsg "Ips client $client server $server"
+	# Start tcpdump
+	LogMsg "Starting tcpdump on $client"
+	ssh ${client} "sh -c 'timeout 10 tcpdump -i eth1 icmp -w tcpdumptest_${logSuffix}.pcap &'"
+	ping_test $logSuffix
+}
+
+# This function will remove patch from xdpdump
+function remove_xdpdump_patch() {
+	ssh ${client} "cd bpf-samples && patch -p1 -R < /root/xdpdump_kern_${ACTION}.patch"
+	check_exit_status "Reverting XDPDump patch: /root/xdpdump_kern_${ACTION}.patch"
+}
+
+
+UTIL_FILE="./utils.sh"
+
+# Source utils.sh
+. ${UTIL_FILE} || {
+		echo "ERROR: unable to source ${UTIL_FILE}!"
+		echo "TestAborted" > state.txt
+		exit 0
+}
+
+# Source constants file and initialize most common variables
+UtilsInit
+
+# Script start from here
+LogMsg "*********INFO: Script Execution Started********"
+
+# check for client and server ips are present
+for ip in ${client} ${server}
+do
+	CheckIP ${ip}
+	if [ $? -eq 1 ]; then
+		LogErr "ERROR: Please provide valide client and server ip. Invalide ip: ${ip}"
+		SetTestStateAborted
+		exit 1
```

Comment:
Exit with 1 will cause Run-LinuxCmd to retry if appropriate option is not used. Please check.